Refactor SmsService to use OAuth2 Client Credentials Grant

forkimenjeckayang opened the pull request:
**Description:**
This PR refactors the `SmsService` class to replace Basic Authentication with OAuth2 Client Credentials Grant for interacting with the SMS provider API. Key changes include:
- Added token fetching logic with caching in `getAccessToken`.
- Updated the `ApiClient` interceptor to use `Authorization: Bearer` headers.
- Tested successfully with a Prism mock server and the `ssegning-tw-01` realm's 'Phone number login' flow.

**Main Changes:**
- `SmsService.java`: Implemented OAuth2 logic.
- `sms-open-api.yml`: Added `/token` endpoint and updated security to `oauth2`.
- `compose.yml`: Added `OAUTH2_CLIENT_ID`, `OAUTH2_CLIENT_SECRET`, `OAUTH2_TOKEN_ENDPOINT`.
- `README.md`: Updated with OAuth2 details, Docker-Kubernetes examples, and environment variables.

stephane-segning commented on openapi/sms-open-api.yml:
```diff
@@ -117,6 +171,9 @@ components:
       required:
         - status
   securitySchemes:
-    basic_auth:
```
Don't remove basic auth

stephane-segning commented on openapi/sms-open-api.yml:
```diff
@@ -69,6 +69,60 @@ paths:
               schema:
                 $ref: "#/components/schemas/ConfirmTanResponse"
 
+  /token:  # New endpoint for OAuth2 token issuance
```
I don't think we need a token endpoint in the OpenAPI File

stephane-segning commented on src/main/java/com/vymalo/keycloak/authenticator/PhoneNumberSendTan.java:
```diff
@@ -26,9 +26,8 @@ public void authenticate(AuthenticationFlowContext context) {
 
         if (hash$.isPresent()) {
             final var hash = hash$.get();
-            event.clone()
```
Why the formatting change?

forkimenjeckayang replied:
Due to this error: The method clone() from the type Object is not visibleJava(67108965)
Object java.lang.Object.clone() throws CloneNotSupportedException

stephane-segning commented on src/main/java/com/vymalo/keycloak/authenticator/PhoneNumberValidateTan.java:
```diff
@@ -64,8 +64,7 @@ public void action(AuthenticationFlowContext context) {
             context.success();
         } else {
             final var event = context.getEvent();
-            event.clone()
```
Same

forkimenjeckayang replied:
Same thing: The method clone() from the type Object is not visibleJava(67108965)
Object java.lang.Object.clone() throws CloneNotSupportedException

stephane-segning commented on src/main/java/com/vymalo/keycloak/services/SmsService.java:
```diff
@@ -10,39 +10,55 @@
         final var apiClient = new ApiClient();
         apiClient.updateBaseUri(smsUrl);
         apiClient.setRequestInterceptor(builder -> {
-            if (StringUtils.isNotEmpty(basicUsr) && StringUtils.isNotEmpty(basicPwd)) {
```
Don't remove Basic auth, but add OAuth2 too.

stephane-segning replied:
OAuth2 shall be added only if corresponding variables are present, i.e `OAUTH_TOKEN_ENDPOINT`

stephane-segning commented on README.md:
```diff
@@ -110,10 +117,12 @@ spec:
               value: "http://your-sms-api-url"
             - name: SMS_API_COUNTRY_PATTERN
               value: "cm|de|fr"
-            - name: SMS_API_AUTH_USERNAME
```
Don't remove Basic Auth

stephane-segning commented on src/main/java/com/vymalo/keycloak/services/SmsService.java:
```diff
@@ -10,39 +10,55 @@
+
     static {
         final var smsUrl = Utils.getEnv(ConfigKey.CONF_PRP_SMS_URL);
-        final var basicUsr = Utils.getEnv(ConfigKey.BASIC_AUTH_USERNAME);
```
Same

stephane-segning reviewed (changes requested):
Good work! Please review these small points

stephane-segning reviewed (commented):
I guess this code is working. This shall now be tested with various versions of Keycloak (from 22 to 26) to ensure it works perfectly

forkimenjeckayang commented:
Hello Sir @stephane-segning ,I’ve addressed your comment by testing the plugin across Keycloak versions 22 to 26 to ensure compatibility.

To support this, I updated the `docker-compose.yml` to include `depends_on` with `service_healthy` conditions for `token-mock` across all Keycloak services (`keycloak-22` to `keycloak-26`). This ensures the mock OAuth2 token endpoint is ready before Keycloak starts.

Here’s what I did:
- **WireMock Setup**: Added `health.json` alongside `token-mock.json` in `./wiremock/mappings/` to handle `GET /health` and `POST /token`, fixing the `token-mock` health check.
- **Docker Compose**: Standardized the config for all Keycloak versions to use the same mock services and environment variables (e.g., `OAUTH2_TOKEN_ENDPOINT=http://token-mock:8080/token`).
- **Testing**: Ran `docker-compose up keycloak-{KC-Version} prism token-mock` and confirmed the SMS flow works perfectly with all versions (22–26), with `prism` returning `200` for `POST /sms/send` in each case.

stephane-segning reviewed (approved):
Perfect work. Thanks
